source/opt/dead_branch_elim_pass.cpp

```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
+          &caseVal,
+          &selVal,
+          &liveLabId,
+        ](const uint32_t* idp) {
+          if (icnt == 1) {
+            // Start with default label.
+            liveLabId = *idp;
+          } else if (icnt > 1) {
+            if (icnt % 2 == 0) {
+              caseVal = *idp;
+            } else {
+              if (caseVal == selVal) liveLabId = *idp;
             }
-            ++icnt;
-          });
+          }
+          ++icnt;
+        });
+      }
     }
 
-    const uint32_t mergeLabId =
-        mergeInst->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx);
-    AddBranch(liveLabId, *bi);
-    backedges_.erase(br);
-    context()->KillInst(br);
-    context()->KillInst(mergeInst);
-
-    modified = true;
-
-    // Iterate to merge block adding dead blocks to elimination set
-    auto dbi = bi;
-    ++dbi;
-    uint32_t dLabId = (*dbi)->id();
-    while (dLabId != mergeLabId) {
-      if (!HasNonPhiNonBackedgeRef(dLabId)) {
-        // Kill use/def for all instructions and mark block for elimination
-        backedges_.erase((*dbi)->terminator());
-        KillAllInsts(*dbi);
-        elimBlocks.insert(*dbi);
+    // Don't simplify branches of continue blocks. A path from the continue to
+    // the merge is required.
+    // TODO(alan-baker): They can be simplified iff there remains a path to the
+    // backedge. Structured control flow should guarantee one path hits the
+    // backedge, but I've removed the requirement for structured control flow
+    // from this pass.
+    bool simplify = liveLabId != 0 && !continues.count(block);
+
+    if (simplify) {
+      modified = true;
+      // Replace with unconditional branch.
+      // Remove the merge instruction if it is a selection merge.
+      AddBranch(liveLabId, block);
+      context()->KillInst(terminator);
+      ir::Instruction* mergeInst = block->GetMergeInst();
+      if (mergeInst->opcode() == SpvOpSelectionMerge) {
+        context()->KillInst(mergeInst);
       }
-      ++dbi;
-      dLabId = (*dbi)->id();
+      stack.push_back(get_parent_block(liveLabId));
+    } else {
+      // All successors are live.
+      block->ForEachSuccessorLabel(
+          [&stack, get_parent_block](const uint32_t label) {
+            stack.push_back(get_parent_block(label));
+          });
     }
+  }
 
-    // If merge block is unreachable, continue eliminating blocks until
-    // a live block or last block is reached.
-    while (!HasNonPhiNonBackedgeRef(dLabId)) {
-      backedges_.erase((*dbi)->terminator());
-      KillAllInsts(*dbi);
-      elimBlocks.insert(*dbi);
-      ++dbi;
-      if (dbi == structuredOrder.end()) break;
-      dLabId = (*dbi)->id();
+  // Check for unreachable merge and continue blocks with live headers, those
+  // blocks must remain. Continues are tracked separately so that when updating
+  // live phi nodes with an edge from a continue I can replace it with an
+  // undef (because I still clobber the continue block).
```

Comment:
"(because we clobber the instructions inside of the continue block)."  

Most important is the clarification that the definition instructions inside the continue block are removed, but the continue block itself survives.

The "still" wasn't doing anything for you here.  And I use the royal "we" because the reader should feel they own the code too.

Reply:
Updated